core/chains/evm/logpoller/orm.go

```diff
@@ -313,34 +314,29 @@ type Exp struct {
 	ShouldDelete bool
 }
 
+// DeleteExpiredLogs removes any logs which either:
+//   - don't match any currently registered filters, or
+//   - have a timestamp older than any matching filter's retention, UNLESS there is at
+//     least one matching filter with retention=0
 func (o *DSORM) DeleteExpiredLogs(ctx context.Context, limit int64) (int64, error) {
```

Comment:
Is it possible to accidentally remove logs when job is updated? Let's imagine the following scenario 

* We have a job, that registered multiple LogPoller filters during the init
* CLO proposes a job replacement, a slight change to the job definition, but contracts stay the same so filters would be the same, so the logs required for plugin to work
* NOP accepts the job which triggers removing the old job and creating a new one. It includes unregistering old filters and registering new ones
   * In the meantime LogPoller `PruneExpiredLogs` kicks in trying to remove stale logs. It sees a new set of "abandoned" logs, because filters were just removed. It removes these logs immediately.
* Plugin init registers new filters (which are the same), but logs are gone so we might need to run `replay`

I wonder if this is something that could happen or impact the system. How does it work for other products? What if job is removed and added in a two way process? (via API). AFAIK CLO does this entire flow in a large single db transaction, but I guess it might change at some point because of the loopps architecture, right?

Reply:
Sounds good. One thing that comes to my mind that might work as an additional safety check is to wait for some "sane" threshold before deleting logs without the filter. When `retention is null` we match only logs older than X days. Not sure what that threshold should be (maybe configurable?), but that might prevent us from accidentally losing logs right after deleting the job in some weird way. WDYT?

Reply:
Yes, I was thinking that might be what we have to do, if we can't solve things with ChainReader.

Reply:
Actually, I just thought of a simple solution which I think would be pretty robust:  in `UnregisterFilter()`, we could return successful but delay the actual removal of the filter by 1 hour (or any long enough time, it could even be 24 hours but that seems like overkill).  As long as any call to` RegisterFilter()` arrives within that 1 hour grace period that matches the logs that would have been deleted, there is no time at which the pruner would be allowed to delete them.